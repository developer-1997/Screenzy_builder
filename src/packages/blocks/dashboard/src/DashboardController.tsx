import { IBlock } from "../../../framework/src/IBlock";
import { Message } from "../../../framework/src/Message";
import { BlockComponent } from "../../../framework/src/BlockComponent";
import MessageEnum, {
  getName
} from "../../../framework/src/Messages/MessageEnum";
import { runEngine } from "../../../framework/src/RunEngine";

// Customizable Area Start
// @ts-ignore
import RNAdvertisingId from "react-native-advertising-id";
import Toast from "react-native-simple-toast";
import { createRef } from "react";
// Customizable Area End

export const configJSON = require("./config.js");

export interface Props {
  navigation: any;
  id: string;
  // Customizable Area Start
  // Customizable Area End
}
interface S {
  // Customizable Area Start
  dashboardData: any;
  token: string;
  errorMsg: string;
  loading: boolean;
  count: number;
  quizeData: any;
  sound: any;
  quizQuestionLoaded: boolean;
  quizQuestionData: any;
  events: any;
  offline: boolean;
  offlineVideo: any;
  quizid: any;
  userName: string;
  userMobile: string;
  userEmail: string;
  videoPlay: boolean;
  video: any;
  counter: number;
  videoLoader: boolean;
  videoPlayerKey?: string | number;
  bubbleHide: boolean;
  companyId: any;
  videoId: any;
  locationPermissionGranted: boolean;
  bubble_time: any;
  dynamic_quiz: any;
  cacheVideoURL: string | null;
  dynamicVideoURL: string;
  last_dynamic_video: any;
  staticVideoCurrentTime: number | string;
  static_adv: boolean;
  geofence: string;
  cacheVideoID: string | number;
  resetSeek: boolean;
  correct_Answer: any;
  staticVideo: any;
  staticPause: boolean;
  allStaticVideos: any;
  allDynamicVideos: any;
  allGeofenes: any;
  cameraOn: any;
  allVideos1: any;
  // Customizable Area End
}
interface SS {
  id: any;
}

export default class DashboardController extends BlockComponent<Props, S, SS> {
  // Customizable Area Start
  apiDashboardItemCallId: string = "";
  quizSaveAnswerId: any = "";
  dynamicStatsId: any = "";
  dashboardApiCallId: string = "";
  apiGetQueryStrinurl: string = "";
  getQuizeDataApiCallId: any = "";
  getQuizesDataApiCallId: any = "";
  unsubscribe: any = null;
  userDataApiCallId: any = "";
  userVideoDataId: any = "";
  lastNumber: number = -1;
  offlineVideoDataId: any = "";
  lat: any = "";
  lng: any = "";
  getBubbleTimerId: any = "";
  getAllVideos1Id: any = "";
  videoCurrentTime: any;
  bubbleUpTime: any;
  isFirstTimeOnline: any;
  stopped: any;
  facesDetected: number;
  totalFacesDetected: any;
  noFacesDetected: any;
  facialCountDataId: any = "";
  tabDataStatusId: any = "";
  facialCountFreq: any;
  offlineVideoNumber: number;
  offlineDynamicNumber: number;
  newFormatQuizDataId: any = "";
  video_Id: any = "";
  company_Id: any = "";
  intervalID: any;
  timeupate: any;
  timeOut: any;
  questionTime: number;
  transLat: any;
  transLong: any;
  cameraTimer: any;
  // Customizable Area End
  
  constructor(props: Props) {
    super(props);
    this.receive = this.receive.bind(this);
    console.disableYellowBox = true;
    // Customizable Area Start
    this.transLat = createRef();
    this.transLong = createRef();
    this.lat = createRef();
    this.lng = createRef();
    this.videoCurrentTime = createRef();
    this.isFirstTimeOnline = createRef();
    this.stopped = createRef();
    this.totalFacesDetected = createRef();
    this.facesDetected = 0;
    this.noFacesDetected = 0;
    this.facialCountFreq = {};
    this.offlineVideoNumber = 0;
    this.video_Id = createRef();
    this.company_Id = createRef();
    this.timeOut = createRef();
    this.intervalID = 0;
    this.timeupate = null;
    this.offlineDynamicNumber = 0;
    this.questionTime = 5;
    this.subScribedMessages = [
      getName(MessageEnum.AccoutLoginSuccess),
      getName(MessageEnum.RestAPIResponceMessage),
      getName(MessageEnum.SessionSaveMessage),
      getName(MessageEnum.SessionResponseMessage),
      getName(MessageEnum.LocationUpdate),
      getName(MessageEnum.GeoFencingEvent)
    ];

    this.state = {
      dashboardData: [],
      errorMsg: "",
      token: "",
      offlineVideo: null,
      quizid: "",
      userName: "",
      userEmail: "",
      userMobile: "",
      videoPlay: false,
      videoLoader: false,
      videoPlayerKey: 2342,
      bubbleHide: false,
      loading: false,
      count: 0,
      quizeData: "",
      sound: false,
      quizQuestionLoaded: false,
      quizQuestionData: null,
      events: [],
      offline: false,
      video: "",
      counter: 0,
      companyId: "",
      videoId: 0,
      locationPermissionGranted: false,
      bubble_time: 30000,
      dynamic_quiz: null,
      cacheVideoURL: null,
      dynamicVideoURL: "",
      last_dynamic_video: false,
      staticVideoCurrentTime: 0,
      static_adv: false,
      cacheVideoID: 1,
      resetSeek: true,
      geofence: "not touhed",
      correct_Answer: [],
      staticVideo: "",
      staticPause: false,
      allStaticVideos: [],
      allDynamicVideos: [],
      allGeofenes: [],
      cameraOn: false,
      allVideos1: []
    };
    // Customizable Area End
    runEngine.attachBuildingBlock(this as IBlock, this.subScribedMessages);
  }

  async componentDidMount() {
    super.componentDidMount();
    this.getToken();
    if (this.isPlatformWeb() === false) {
      this.props.navigation.addListener('willFocus', () => {
        this.getToken();
      });
    }
    // Customizable Area Start
    // Customizable Area End
  }
  
  getToken=()=>{
    const msg: Message = new Message(getName(MessageEnum.SessionRequestMessage));
    this.send(msg);
  }

  getDashboardData(): boolean {
    // Customizable Area Start
    const header = {
      "Content-Type": configJSON.dashboarContentType,
      token: this.state.token
    };
    const requestMessage = new Message(
      getName(MessageEnum.RestAPIRequestMessage)
    );
    this.apiDashboardItemCallId = requestMessage.messageId;
    requestMessage.addData(
      getName(MessageEnum.RestAPIResponceEndPointMessage),
      configJSON.dashboardGetUrl
    );

    requestMessage.addData(
      getName(MessageEnum.RestAPIRequestHeaderMessage),
      JSON.stringify(header)
    );

    requestMessage.addData(
      getName(MessageEnum.RestAPIRequestMethodMessage),
      configJSON.dashboarApiMethodType
    );
    runEngine.sendMessage(requestMessage.id, requestMessage);
    // Customizable Area End
    return true;
  }

  async receive(from: string, message: Message) {
    // Customizable Area Start
    if (
      this.quizSaveAnswerId ===
      message.getData(getName(MessageEnum.RestAPIResponceDataMessage))
    ) {
      const apiResponse = message.getData(
        getName(MessageEnum.RestAPIResponceSuccessMessage)
      );
      console.log("api response from the api", apiResponse);
    }

    if (
      this.newFormatQuizDataId ===
      message.getData(getName(MessageEnum.RestAPIResponceDataMessage))
    ) {
      const apiResponse = message.getData(
        getName(MessageEnum.RestAPIResponceSuccessMessage)
      );
      this.setState({ dynamic_quiz: apiResponse.data.quiz_id });
      let NavigateScreen = apiResponse.data.quiz_type;
      if (NavigateScreen) {
        this.setState({ quizid: apiResponse.data.quiz_id });
        this.setState({
          quizQuestionLoaded: false,
          quizQuestionData: null
        });
        this.props.navigation.navigate(this.GetQuizTypeName(NavigateScreen), {
          quizId: apiResponse.data.quiz_id,
          theme: apiResponse.data.theme,
          question_Time: apiResponse.data.question_time
        });
      }
    }
    if (
      this.facialCountDataId ===
      message.getData(getName(MessageEnum.RestAPIResponceDataMessage))
    ) {
      const apiResponse = message.getData(
        getName(MessageEnum.RestAPIResponceSuccessMessage)
      );
      console.log(apiResponse);
      this.facesDetected = 0;
    }

    if (
      this.getBubbleTimerId ===
      message.getData(getName(MessageEnum.RestAPIResponceDataMessage))
    ) {
      const apiResponse = message.getData(
        getName(MessageEnum.RestAPIResponceSuccessMessage)
      );
      this.setState({ bubble_time: apiResponse.bubble_time });
    }
    if (
      this.getQuizeDataApiCallId ===
      message.getData(getName(MessageEnum.RestAPIResponceDataMessage))
    ) {
      const apiResponse = message.getData(
        getName(MessageEnum.RestAPIResponceSuccessMessage)
      );

      this.setState({ quizeData: apiResponse });
      this.setState({
        quizQuestionLoaded: true,
        quizQuestionData: apiResponse.data
      });
    }
    if (
      this.getQuizesDataApiCallId ===
      message.getData(getName(MessageEnum.RestAPIResponceDataMessage))
    ) {
      const apiResponse = message.getData(
        getName(MessageEnum.RestAPIResponceSuccessMessage)
      );

      if (Array.isArray(apiResponse.data)) {
        let RandomIndex = Math.floor(Math.random() * apiResponse.data.length);

        let NavigateScreen = apiResponse.data[RandomIndex].attributes.quiz_type;
        if (NavigateScreen) {
          this.setState({ quizid: apiResponse.data[RandomIndex].id });
          this.setState({
            quizQuestionLoaded: false,
            quizQuestionData: null
          });
          this.props.navigation.navigate(this.GetQuizTypeName(NavigateScreen), {
            quizId: apiResponse.data[RandomIndex].id,
            theme: apiResponse.data[RandomIndex].attributes.theme
          });
        }
      }
    }
    if (getName(MessageEnum.RestAPIResponceMessage) === message.id) {
      let responseJson = message.getData(
        getName(MessageEnum.RestAPIResponceSuccessMessage)
      );
      if (responseJson && !responseJson.errors && responseJson.data) {
        if (responseJson.data.length === 0) {
          this.setState({
            errorMsg: "Data Not Found",
            loading: false
          });
        } else {
          this.setState({
            dashboardData: responseJson.data,
            errorMsg: "",
            loading: false
          });
        }
      } else {
        let errorReponse = message.getData(
          getName(MessageEnum.RestAPIResponceErrorMessage)
        );
        if (errorReponse === undefined) {
          this.setState({
            errorMsg: "Something went serioulsy wrong",
            loading: false
          });
        } else {
          this.setState({
            errorMsg: errorReponse,
            loading: false
          });
        }
      }
    }
    if (
      this.userDataApiCallId ===
      message.getData(getName(MessageEnum.RestAPIResponceDataMessage))
    ) {
      const apiResponse = message.getData(
        getName(MessageEnum.RestAPIResponceSuccessMessage)
      );
      console.log(apiResponse);
    }
    // Customizable Area End
  }

  // Customizable Area Start
  sendHelloMessage() {
    // console.log("hello  send Hello Message")
    const msg: Message = new Message(getName(MessageEnum.StartGeoFencing));
    msg.properties.text = "hello world";
    this.send(msg);
  }
  GetQuizTypeName = (type: string) => {
    switch (type) {
      case "Quiz-1":
        return "Quiz1";
      case "Quiz-2":
        return "Quiz1";
      case "Quiz-3":
        return "Quiz1";
      case "Quiz-4":
        return "Quiz1";
      case "Quiz-5":
        return "Quiz2";
      case "Quiz-6":
        return "Quiz2";
      case "Quiz-7":
        return "Quiz6";
      case "Quiz-8":
        return "Quiz6";
      case "Quiz-12":
        return "Quiz6";
      case "Quiz-9":
        return "Quiz5";
      case "Quiz-10":
        return "Quiz5";
      case "Quiz-11":
        return "Quiz5";
      case "Quiz-14":
        return "Quiz4";
      case "Quiz-13":
        return "Quiz3";

      default:
        return null;
    }
  };
  fetchQuizeData = (id: any) => {
    const headers = {
      "Content-Type": configJSON.dashboarContentType,
      token: configJSON.staticToken
    };
    const getQuizeDataValidation = new Message(
      getName(MessageEnum.RestAPIRequestMessage)
    );
    this.getQuizeDataApiCallId = getQuizeDataValidation.messageId;

    getQuizeDataValidation.addData(
      getName(MessageEnum.RestAPIResponceEndPointMessage),
      "bx_block_cfquiztrivia/quizzes/" + id
    );
    getQuizeDataValidation.addData(
      getName(MessageEnum.RestAPIRequestHeaderMessage),
      JSON.stringify(headers)
    );

    getQuizeDataValidation.addData(
      getName(MessageEnum.RestAPIRequestMethodMessage),
      "GET"
    );

    runEngine.sendMessage(getQuizeDataValidation.id, getQuizeDataValidation);
  };
  fetchallQuizData = () => {
    const headers = {
      "Content-Type": configJSON.dashboarContentType,
      token: configJSON.staticToken
    };
    const getQuizesDataValidation = new Message(
      getName(MessageEnum.RestAPIRequestMessage)
    );
    this.getQuizesDataApiCallId = getQuizesDataValidation.messageId;

    getQuizesDataValidation.addData(
      getName(MessageEnum.RestAPIResponceEndPointMessage),
      "bx_block_cfquiztrivia/quizzes"
    );

    getQuizesDataValidation.addData(
      getName(MessageEnum.RestAPIRequestHeaderMessage),
      JSON.stringify(headers)
    );

    getQuizesDataValidation.addData(
      getName(MessageEnum.RestAPIRequestMethodMessage),
      "GET"
    );

    runEngine.sendMessage(getQuizesDataValidation.id, getQuizesDataValidation);
  };
  sendUserData = async (id: any) => {
    let { advertisingId } = await RNAdvertisingId.getAdvertisingId();
    const UserDataValidation = new Message(
      getName(MessageEnum.RestAPIRequestMessage)
    );
    this.userDataApiCallId = UserDataValidation.messageId;

    UserDataValidation.addData(
      getName(MessageEnum.RestAPIResponceEndPointMessage),
      "bx_block_cfquiztrivia/participents"
    );

    const header = {
      "Content-Type": "multipart/form-data",
      token: configJSON.staticToken
    };
    UserDataValidation.addData(
      getName(MessageEnum.RestAPIRequestHeaderMessage),
      JSON.stringify(header)
    );
    UserDataValidation.addData(
      getName(MessageEnum.RestAPIRequestMethodMessage),
      "POST"
    );

    let formdata = new FormData();
    formdata.append("name", this.state.userName);
    formdata.append("email", this.state.userEmail);
    formdata.append("full_phone_number", this.state.userMobile);
    formdata.append("quiz_id", id);
    formdata.append("unique_serial_no", advertisingId);

    UserDataValidation.addData(
      getName(MessageEnum.RestAPIRequestBodyMessage),
      formdata
    );
    runEngine.sendMessage(UserDataValidation.id, UserDataValidation);
    console.log("quiz idf----from api hit", id);

    if (
      this.state.userEmail.length > 0 ||
      this.state.userMobile.length > 0 ||
      this.state.userName.length > 0
    ) {
      Toast.show("Your details are successfully submitted.", Toast.SHORT);
      this.setState({ userName: "", userMobile: "", userEmail: "" });
      this.props.navigation.navigate("Home");
    } else {
      this.setState({ userName: "", userMobile: "", userEmail: "" });
      this.props.navigation.navigate("Home");
    }
  };

  updateTimer = () => {
    const header = {
      "Content-Type": configJSON.dashboarContentType,
      token: configJSON.staticToken
    };
    const getBubbleDataValidation = new Message(
      getName(MessageEnum.RestAPIRequestMessage)
    );
    this.getBubbleTimerId = getBubbleDataValidation.messageId;

    getBubbleDataValidation.addData(
      getName(MessageEnum.RestAPIResponceEndPointMessage),
      "bx_block_cfquiztrivia/get_bubble_time"
    );
    getBubbleDataValidation.addData(
      getName(MessageEnum.RestAPIRequestHeaderMessage),
      JSON.stringify(header)
    );

    getBubbleDataValidation.addData(
      getName(MessageEnum.RestAPIRequestMethodMessage),
      "GET"
    );

    runEngine.sendMessage(getBubbleDataValidation.id, getBubbleDataValidation);
  };
  locationBasedQuiz = () => {
    const headers = {
      "Content-Type": configJSON.dashboarContentType,
      token: configJSON.staticToken
    };

    const Quizdata = new Message(getName(MessageEnum.RestAPIRequestMessage));
    this.newFormatQuizDataId = Quizdata.messageId;

    Quizdata.addData(
      getName(MessageEnum.RestAPIRequestHeaderMessage),
      JSON.stringify(headers)
    );
    if (this.state.dynamic_quiz == null) {
      Quizdata.addData(
        getName(MessageEnum.RestAPIResponceEndPointMessage),
        `bx_block_cfquiztrivia/sequence_wise_quiz?quiz_id=`
      );
    } else {
      Quizdata.addData(
        getName(MessageEnum.RestAPIResponceEndPointMessage),
        `bx_block_cfquiztrivia/sequence_wise_quiz?quiz_id=${
          this.state.dynamic_quiz
        }`
      );
    }

    Quizdata.addData(getName(MessageEnum.RestAPIRequestMethodMessage), "GET");

    runEngine.sendMessage(Quizdata.id, Quizdata);
  };
  saveUserQuizAnswers = async (id: any, answers: any) => {
    let { advertisingId } = await RNAdvertisingId.getAdvertisingId();
    const headers = {
      "Content-Type": configJSON.dashboarContentType,
      token: configJSON.staticToken
    };
    let body: any = {};
    body.quiz_id = id;
    body.user_answers = answers;
    body.unique_serial_no = advertisingId;

    const QuizAnswerData = new Message(
      getName(MessageEnum.RestAPIRequestMessage)
    );

    this.quizSaveAnswerId = QuizAnswerData.messageId;

    QuizAnswerData.addData(
      getName(MessageEnum.RestAPIResponceEndPointMessage),
      "bx_block_cfquiztrivia/quizzes/save_answers"
    );

    QuizAnswerData.addData(
      getName(MessageEnum.RestAPIRequestBodyMessage),
      JSON.stringify(body)
    );
    QuizAnswerData.addData(
      getName(MessageEnum.RestAPIRequestHeaderMessage),
      JSON.stringify(headers)
    );
    QuizAnswerData.addData(
      getName(MessageEnum.RestAPIRequestMethodMessage),
      "POST"
    );
    runEngine.sendMessage(QuizAnswerData.id, QuizAnswerData);
  };
  // Customizable Area End

}
