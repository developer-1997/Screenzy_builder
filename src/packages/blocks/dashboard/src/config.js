Object.defineProperty(exports, "__esModule", {
  value: true
});

// Customizable Area Start
exports.dashboardGetUrl = "/dashboard/dashboards";
exports.dashboarContentType = "application/json";
exports.dashboarApiMethodType = "GET";
exports.dashboardHost = "<calculated when request is sent>";
exports.dashboarUserAgent = "PostmanRuntime/7.26.5";
exports.dashboarAccept = "*/*";
exports.dashboarAcceptEncoding = "gzip, deflate, br";
exports.dashboarConnection = "keep-alive";
exports.dashboartoken = "";
exports.labelTitleText = "dashboard";
exports.labelBodyText = "dashboard Body";
exports.liveVideoURL =  "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerFun.mp4";
exports.offlineVideoURL = "https://samplelib.com/lib/preview/mp4/sample-10s.mp4";
exports.dummyofflineVideoURL = "https://samplelib.com/lib/preview/mp4/sample-5s.mp4";
exports.staticToken = "tyugiojpkiy8ucfghvjbuy76tfygui86t7yuhu7vbnubhhjihbjihjghjkiou897ytfyghjkiou8976tfygvhjbkiu8y76t5tfychjbuhy"
//Customizable Area End