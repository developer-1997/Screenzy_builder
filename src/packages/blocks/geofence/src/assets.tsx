export const AddIcon = require("../assets/add.svg") 