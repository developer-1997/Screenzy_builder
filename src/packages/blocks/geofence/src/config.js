Object.defineProperty(exports, "__esModule", {
  value: true
});

// Customizable Area Start
exports.labelTitleText = "Geofence";
exports.labelBodyText = "Geofence Body";
exports.radarKey = "prj_live_pk_dc7bbaa26e61343e20a955965be95a7035dd611a";
exports.geofenceListApi="/bx_block_geofence2/geofence" ;
exports.geofenceListMethod="GET"
exports.geofenceContentType="application/json";
exports.geofenceListTitle="List of geofences"
exports.geofenceAddTitle="Add"
exports.editTitle="Editor"
exports.backButtonTitle="Back"
exports.geofenceDeleteMethodType="DELETE";
exports.geofenceUpdateMethodType="PUT"
exports.geofenceCreateMethodType="POST"
// Customizable Area End