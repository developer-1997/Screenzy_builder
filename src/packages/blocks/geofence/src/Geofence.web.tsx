import React from "react";

// Customizable Area Start
import './geofence.css'
// Customizable Area End

// Customizable Area Start
import GeofenceList from "../../../components/src/GeofenceList.web";
import Map from "../../../components/src/Map.web";
import GeofenceEditor from "../../../components/src/GeofenceEditor.web";
import GeofenceBackButton from "../../../components/src/GeofenceBackButton.web";
import { AddIcon } from "./assets"
import GeofenceAddButton from "../../../components/src/GeofenceAddButton.web";
// Customizable Area End

import GeofenceController, {
  Props,
  configJSON,
} from "./GeofenceController";

export default class Geofence extends GeofenceController {
  constructor(props: Props) {
    super(props);
    // Customizable Area Start
    // Customizable Area End
  }

  // Customizable Area Start
  
  // Customizable Area End

  render() {
    return (
      // Customizable Area Start
      <>
        <div className="container">
          <div className="controls">

          {this.state.editMode === "VIEW" ? (
            <>
              
              <GeofenceList 
                data-test-id="geofenceListId"
                title={configJSON.geofenceListTitle}
                geofenceList={this.state.geofenceList}
                handleEditClick={this.handleEditClick}
              />


              <GeofenceAddButton 
                data-test-id="addButtonId"
                AddTitle={configJSON.geofenceAddTitle}
                AddIcon={AddIcon}
                createGeofenceHandler={this.createGeofenceHandler}
              />
            </>
          ) : (
            <>
              <GeofenceBackButton 
                data-test-id="backButtonId"
                handleBackClick={this.handleBackClick}
                backButtonTitle={configJSON.backButtonTitle}
              />
              <GeofenceEditor 
                data-test-id="geofenceEditor"
                editMode={this.state.editMode}
                editableCircle={this.state.editableCircle}
                lat={this.state.lat}
                lng={this.state.lng}
                rad={this.state.rad}
                idf={this.state.idf}
                handlerIdfChange={this.handlerIdfChange}
                handlerRadChange={this.handlerRadChange}
                handlerLatChange={this.handlerLatChange}
                handlerLngChange={this.handlerLngChange}
                HandleSubmit={this.HandleSubmit}
                disableHandler={this.disableHandler}
                HandleDelete={this.HandleDelete}
                editTitle={configJSON.editTitle}
              />
            </>
          )}
          </div>
          <div className="map">
            <Map
              geofences={this.state.geofenceList}
              editMode={this.state.editMode}
              mapCenter={this.state.mapCenter}
              editableCircle={this.state.editableCircle}
              handleSetEditableCircle={this.handleSetEditableCircle}
            />
          </div>
        </div>
      </>
      // Customizable Area End
    );
  }
}

// Customizable Area Start
// Customizable Area End
