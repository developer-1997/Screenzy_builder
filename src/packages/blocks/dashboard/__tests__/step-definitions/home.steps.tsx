import React from "react";
import { defineFeature, loadFeature } from "jest-cucumber";
import { shallow, ShallowWrapper } from "enzyme";
import renderer from "react-test-renderer";
import * as helpers from "../../../../framework/src/Helpers";
import axios from "axios";
import mockRNCNetInfo from "@react-native-community/netinfo/jest/netinfo-mock.js";
import mockRNDeviceInfo from "react-native-device-info/jest/react-native-device-info-mock";

const headers = {
  token: "tyugiojpkiy8ucfghvjbuy76tfygui86t7yuhu7vbnubhhjihbjihjghjkiou897ytfyghjkiou8976tfygvhjbkiu8y76t5tfychjbuhy",
};

import RNFetchBlob from "rn-fetch-blob";
import RNFS from "react-native-fs";

import Home from "../../src/Home";

const navigation = require("react-navigation");
const screenProps = {
  navigation: navigation,
  id: "Home",
};
let homeWrapper: ShallowWrapper;
let instance: Home;

const feature = loadFeature("./__tests__/features/home-scenario.feature");
jest.mock("@react-native-community/netinfo", () => mockRNCNetInfo);
jest.mock("react-native-device-info", () => mockRNDeviceInfo);

const baseURL =
  "https://screenzycommercialsllp-125263-ruby.b125263.dev.eastus.az.svc.builder.cafe";

const mockFiles = [
  {
    path: "/storage/emulated/0/Android/data/com.ScreenzyCommercialsLLP/Screenzy/video-687-video-dynamic-22.55458-88.34081-.mp4",
  },
  {
    path: "/storage/emulated/0/Android/data/com.ScreenzyCommercialsLLP/Screenzy/video-687-video-dynamic-22.55458-88.34081-.mp4",
  },
  {
    path: "/storage/emulated/0/Android/data/com.ScreenzyCommercialsLLP/Screenzy/video-687-video-dynamic-22.55458-88.34081-.mp4",
  },
];

jest.mock(
  "react-native-advertising-id",
  () => {
    return {
      RNAdvertisingId: jest.fn().mockImplementation(() => Promise.resolve()),
    };
  },
  { virtual: true }
);

jest.mock("react-native/Libraries/EventEmitter/NativeEventEmitter");
jest.mock("react-native-simple-toast", () => ({
  SHORT: jest.fn(),
}));
jest.mock("rn-fetch-blob", () => ({
  fs: {
    lstat: jest.fn(),
    dirs: {
      SDCardApplicationDir: "/Screenzy",
    },
  },
}));
jest.mock("axios");
jest.mock("react-native-fs", () => ({
  readDir: jest.fn().mockResolvedValue([
    {
      path: "/storage/emulated/0/Android/data/com.ScreenzyCommercialsLLP/Screenzy/video-687-video-dynamic-22.55458-88.34081-.mp4",
    },
    {
      path: "/storage/emulated/0/Android/data/com.ScreenzyCommercialsLLP/Screenzy/video-687-video-dynamic-22.55458-88.34081-.mp4",
    },
    {
      path: "/storage/emulated/0/Android/data/com.ScreenzyCommercialsLLP/Screenzy/video-687-video-dynamic-22.55458-88.34081-.mp4",
    },
  ]),
}));
let response: any;

defineFeature(feature, (test) => {
  beforeEach(() => {
    jest.resetModules();
    jest.doMock("react-native", () => ({ Platform: { OS: "web" } }));
    jest.spyOn(helpers, "getOS").mockImplementation(() => "web");
  });
  test("User navigates to home", ({ given, when, then }) => {
    given("I am a User loading home", () => {
      homeWrapper = shallow(<Home {...screenProps} />);
      expect(homeWrapper).toBeTruthy();
      expect(homeWrapper).toMatchSnapshot();
    });

    when("I navigate to the home", () => {
      instance = homeWrapper.instance() as Home;
      expect(homeWrapper).toBeTruthy();
      expect(homeWrapper).toMatchSnapshot();
    });

    then("home will load with out errors", async () => {
      expect(homeWrapper).toBeTruthy();
      expect(homeWrapper).toMatchSnapshot();
    });
  });
  test("Get all videos", ({ given, when, then }) => {
    given("I am a User loading home", () => {
      homeWrapper = shallow(<Home {...screenProps} />);
      expect(homeWrapper).toBeTruthy();
      expect(homeWrapper).toMatchSnapshot();
    });
    when("the getAllVideos API endpoint is called", async () => {
      expect(homeWrapper).toBeTruthy();
    });
    then("the response status code should be", async () => {
      try {
        response = await instance.getAllVideos();
        expect(response.status).toBe(200);
      } catch (e) {
        response = e.response;
      }
    });
  });
  test("Geofence data", ({ given, when, then }) => {
    given("The app is launched", () => {
      homeWrapper = shallow(<Home {...screenProps} />);
      expect(homeWrapper).toBeTruthy();
      expect(homeWrapper).toMatchSnapshot();
    });
    when("the geofence data endpoint is called", async () => {
      expect(homeWrapper).toBeTruthy();
    });
    then("should get an array containing geofence data", async () => {
      try {
        response = await instance.getFenceData();
        expect(response.status).toBe(200);
        expect(response).toEqual(
          expect.arrayContaining([
            expect.objectContaining({
              notifyOnEntry: true,
              notifyOnExit: true,
              identifier: expect.any(String),
              latitude: expect.any(Number),
              longitude: expect.any(Number),
              radius: expect.any(Number),
            }),
          ])
        );
      } catch (e) {
        response = e.response;
      }
    });
  });
});

describe("Internal storage mock testing", () => {
  it("returns the correct length of videos in storage", async () => {
    let videosLength = 10;
    RNFetchBlob.fs.lstat.mockResolvedValue(videosLength);
    const result = await instance.videosInStorageLength();
    expect(result).toBe(videosLength);
  });
  it("should get the formatted data from Internal storage", async () => {
    const files = await RNFS.readDir(
      RNFetchBlob.fs.dirs.SDCardApplicationDir + "/Screenzy"
    ).then((filez) => {
      const allPaths = filez.map((i, index) => {
        const videoData = filez.map((j) => {
          const splitArray = j.path.split("-");
          return [splitArray[4], splitArray[1]];
        });
        return {
          path: i.path,
          latitude: videoData[index][0],
          id: videoData[index][1],
        };
      });
      return allPaths;
    });
    expect(files).toEqual([
      {
        path: "/storage/emulated/0/Android/data/com.ScreenzyCommercialsLLP/Screenzy/video-687-video-dynamic-22.55458-88.34081-.mp4",
        latitude: "22.55458",
        id: "687",
      },
      {
        path: "/storage/emulated/0/Android/data/com.ScreenzyCommercialsLLP/Screenzy/video-687-video-dynamic-22.55458-88.34081-.mp4",
        latitude: "22.55458",
        id: "687",
      },
      {
        path: "/storage/emulated/0/Android/data/com.ScreenzyCommercialsLLP/Screenzy/video-687-video-dynamic-22.55458-88.34081-.mp4",
        latitude: "22.55458",
        id: "687",
      },
    ]);
  });
  it("should read and parse files from directory", async () => {
    const spy = jest.spyOn(RNFS, "readDir");
    const expectedPaths = [
      {
        path: "/storage/emulated/0/Android/data/com.ScreenzyCommercialsLLP/Screenzy/video-687-video-dynamic-22.55458-88.34081-.mp4",
        latitude: 22.55458,
        id: "1",
      },
      {
        path: "/storage/emulated/0/Android/data/com.ScreenzyCommercialsLLP/Screenzy/video-687-video-dynamic-22.55458-88.34081-.mp4",
        latitude: 22.55458,
        id: "2",
      },
    ];
    await instance.readPathFromStorage();
    expect(spy).toHaveBeenCalledWith(
      RNFetchBlob.fs.dirs.SDCardApplicationDir + "/Screenzy"
    );
    // verify that the function parses the file paths correctly
    expect(expectedPaths).toEqual([
      { path: mockFiles[0].path, latitude: 22.55458, id: "1" },
      { path: mockFiles[1].path, latitude: 22.55458, id: "2" },
    ]);
  });
});

describe("Post calls testing", () => {
  it("sends expected video stat data to server", async () => {
    const id = 123;
    const unique_serial_no = "abc123";
    const mockResponse = { data: { status: 200, message: "Success" } };
    axios.post.mockResolvedValueOnce(mockResponse);
    await instance.sendStats(id, unique_serial_no);
    expect(axios.post).toHaveBeenCalledTimes(1);
    expect(axios.post).toHaveBeenCalledWith(
      `${baseURL}/advertisments/create_advertisement_report`,
      { advertisement_id: id, unique_serial_no },
      { headers }
    );
    expect(mockResponse.data.status).toEqual(200);
  });

  it("renders with the correct styles", () => {
    const testRenderer = renderer.create(<Home {...screenProps} />);
    const json = testRenderer.toJSON();
    expect(json).toMatchSnapshot();
  });
});