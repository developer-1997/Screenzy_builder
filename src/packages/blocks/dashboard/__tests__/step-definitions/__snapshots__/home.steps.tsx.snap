// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post calls testing renders with the correct styles 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "grey",
        "borderRadius": 39,
        "height": 59,
        "justifyContent": "center",
        "marginTop": 20,
        "opacity": 1,
        "position": "absolute",
        "right": 39,
        "top": 39,
        "width": 59,
        "zIndex": 5,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../packages/blocks/dashboard/assets/sound.png",
        }
      }
      style={
        Object {
          "height": 39,
          "width": 39,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "height": "100%",
        "left": 20,
        "opacity": 0,
        "position": "absolute",
        "top": 100,
        "width": "100%",
        "zIndex": -100,
      }
    }
  />
  <View
    style={
      Object {
        "left": 40,
        "position": "absolute",
        "top": 40,
        "zIndex": 2,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../packages/blocks/dashboard/assets/mainlogo.png",
        }
      }
      style={
        Object {
          "height": 70,
          "width": 140,
        }
      }
    />
  </View>
</View>
`;
